library/src/main/java/com/novoda/downloadmanager/LiteDownloadMigrationService.java

```diff
@@ -81,65 +35,91 @@ public void onCreate() {
-        Intent intent = new Intent(getApplicationContext(), LiteDownloadMigrationService.class);
-        PendingIntent pendingIntent = PendingIntent.getService(this, 1, intent, PendingIntent.FLAG_ONE_SHOT);
-        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
-        if (alarmManager == null) {
-            Log.w(TAG, "Could not retrieve AlarmManager for rescheduling.");
-            return;
-        }
-        alarmManager.set(AlarmManager.RTC_WAKEUP, SystemClock.elapsedRealtime() + 5000, pendingIntent);
+    @Override
+    public void setNotificationChannelCreator(NotificationChannelCreator notificationChannelCreator) {
+        this.notificationChannelCreator = notificationChannelCreator;
     }
 
     @Override
-    public boolean onUnbind(Intent intent) {
-        Log.d(TAG, "onUnbind");
-        migrationCallback = null;
-        Log.d(TAG, "Stopping service");
-        return super.onUnbind(intent);
+    public void setNotificationCreator(NotificationCreator<MigrationStatus> notificationCreator) {
+        this.notificationCreator = notificationCreator;
     }
 
-    class MigrationDownloadServiceBinder extends Binder {
-
-        MigrationDownloadServiceBinder withCallback(Migrator.Callback migrationCallback) {
-            LiteDownloadMigrationService.this.migrationCallback = migrationCallback;
-            return this;
+    @Override
+    public void startMigration() {
+        Optional<NotificationChannel> notificationChannel = notificationChannelCreator.createNotificationChannel();
+        String channelName = notificationChannelCreator.getNotificationChannelName();
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && notificationChannel.isPresent() && notificationChannelDoesNotExist(channelName)) {
+            notificationManager.createNotificationChannel(notificationChannel.get());
         }
 
-        void bind() {
-            if (migrationStatus != null && migrationCallback != null) {
-                LiteDownloadMigrationService.this.migrationCallback.onUpdate(migrationStatus);
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                Migrator migrator = MigrationFactory.createVersionOneToVersionTwoMigrator(
+                        getApplicationContext(),
+                        getDatabasePath("downloads.db"),
+                        migrationStatusCallback
+                );
+                Log.d(TAG, "Begin Migration: " + migrator.getClass());
+                migrator.migrate();
+            }
+        });
+    }
+
+    @TargetApi(Build.VERSION_CODES.O)
+    private boolean notificationChannelDoesNotExist(String channelName) {
+        return notificationManager.getNotificationChannel(channelName) == null;
+    }
+
+    private final Migrator.Callback migrationStatusCallback = new Migrator.Callback() {
+        @Override
+        public void onUpdate(MigrationStatus migrationStatus) {
```

Comment:
We don't pass the `Service` to the `Migrator` anymore. Instead we pass a callback. When updates occur in the `Migrator` we callback here with just the `MigrationStatus` and the `Service` decides how and what to display.